src/Models/HydrostaticFreeSurfaceModels/pcg_implicit_free_surface_solver.jl

```diff
@@ -0,0 +1,151 @@
+
+    return nothing
+end
+
+function compute_implicit_free_surface_right_hand_side!(rhs,
+                                                        implicit_solver::PCGImplicitFreeSurfaceSolver,
+                                                        g, Δt, ∫ᶻQ, η)
+
+    solver = implicit_solver.preconditioned_conjugate_gradient_solver
+    arch = solver.architecture
+    grid = solver.grid
+
+    event = launch!(arch, grid, :xy,
+                    implicit_free_surface_right_hand_side!,
+                    rhs, grid, g, Δt, ∫ᶻQ, η,
+		            dependencies = device_event(arch))
+
+    return event
+end
+
+""" Compute the divergence of fluxes Qu and Qv. """
+@inline flux_div_xyᶜᶜᵃ(i, j, k, grid, Qu, Qv) = δxᶜᵃᵃ(i, j, k, grid, Qu) + δyᵃᶜᵃ(i, j, k, grid, Qv)
+
+@kernel function implicit_free_surface_right_hand_side!(rhs, grid, g, Δt, ∫ᶻQ, η)
+    i, j = @index(Global, NTuple)
+    Az = Azᶜᶜᵃ(i, j, 1, grid)
+    δ_Q = flux_div_xyᶜᶜᵃ(i, j, 1, grid, ∫ᶻQ.u, ∫ᶻQ.v)
+    @inbounds rhs[i, j, 1] = (δ_Q - Az * η[i, j, 1] / Δt) / (g * Δt)
+end
+
+"""
+    implicit_free_surface_linear_operation!(result, x, arch, grid, bcs; args...)
+
+Returns `L(ηⁿ)`, where `ηⁿ` is the free surface displacement at time step `n`
+and `L` is the linear operator that arises
+in an implicit time step for the free surface displacement `η`.
+
+(See the docs section on implicit time stepping.)
+"""
+function implicit_free_surface_linear_operation!(L_ηⁿ⁺¹, ηⁿ⁺¹, ∫ᶻ_Axᶠᶜᶜ, ∫ᶻ_Ayᶜᶠᶜ, g, Δt)
+    grid = L_ηⁿ⁺¹.grid
+    arch = architecture(L_ηⁿ⁺¹)
+
+    fill_halo_regions!(ηⁿ⁺¹, arch)
+
+    event = launch!(arch, grid, :xy, _implicit_free_surface_linear_operation!,
+                    L_ηⁿ⁺¹, grid,  ηⁿ⁺¹, ∫ᶻ_Axᶠᶜᶜ, ∫ᶻ_Ayᶜᶠᶜ, g, Δt,
+                    dependencies = device_event(arch))
+
+    wait(device(arch), event)
+
+    fill_halo_regions!(L_ηⁿ⁺¹, arch)
+
+    return nothing
+end
+
+# Kernels that act on vertically integrated / surface quantities
+@inline ∫ᶻ_Ax_∂x_ηᶠᶜᶜ(i, j, k, grid, ∫ᶻ_Axᶠᶜᶜ, η) = @inbounds ∫ᶻ_Axᶠᶜᶜ[i, j, k] * ∂xᶠᶜᵃ(i, j, k, grid, η)
+@inline ∫ᶻ_Ay_∂y_ηᶜᶠᶜ(i, j, k, grid, ∫ᶻ_Ayᶜᶠᶜ, η) = @inbounds ∫ᶻ_Ayᶜᶠᶜ[i, j, k] * ∂yᶜᶠᵃ(i, j, k, grid, η)
+
+"""
+Compute the horizontal divergence of vertically-uniform quantity using
+vertically-integrated face areas `∫ᶻ_Axᶠᶜᶜ` and `∫ᶻ_Ayᶜᶠᶜ`.
+"""
+@inline Az_∇h²ᶜᶜᵃ(i, j, k, grid, ∫ᶻ_Axᶠᶜᶜ, ∫ᶻ_Ayᶜᶠᶜ, η::ReducedField{X, Y, Nothing}) where {X, Y} =
+    (δxᶜᵃᵃ(i, j, k, grid, ∫ᶻ_Ax_∂x_ηᶠᶜᶜ, ∫ᶻ_Axᶠᶜᶜ, η) +
+     δyᵃᶜᵃ(i, j, k, grid, ∫ᶻ_Ay_∂y_ηᶜᶠᶜ, ∫ᶻ_Ayᶜᶠᶜ, η))
+
+"""
+    _implicit_free_surface_linear_operation!(L_ηⁿ⁺¹, grid, ηⁿ⁺¹, ∫ᶻ_Axᶠᶜᶜ, ∫ᶻ_Ayᶜᶠᶜ, g, Δt)
+
+Returns the left hand side of the "implicit η equation"
+
+```math
+(∇ʰ⋅H∇ʰ - 1/gΔt²) ηⁿ⁺¹ = 1 / (gΔt) ∇ʰ ⋅ Q★ - 1 / (gΔt²) ηⁿ
+----------------------
+        ≡ L_ηⁿ⁺¹
+```
+
+which is written in a discrete finite volume form in which the equation
+is arranged to ensure a symmtric form by multiplying by horizontal areas Az:
```

Comment:
```suggestion
which is written in a discrete finite volume form, that is the equation
is arranged to ensure a symmetric form by multiplying by horizontal areas Az:
```

Reply:
Ah I like the typo fix but the phrase "that is" isn't quite correct, because "discrete finite volume form" is arbitrary and does not imply that it's also symmetric. The symmetry is additional to the fact that the formulation is explicitly discrete.